### ISSUE TITLE
Multiple systemd-activated sockets are attached to wrong rules

### ISSUE THREAD

**Issue by noc0lour** [nixcloud/ip2unix#27]
I'm currently observing with a commandline like: 
```
/usr/local/bin/ip2unix -v -p -r in,tcp,port=8090,systemd=service1 -r in,tcp,port=1113,systemd=service2 -r out,ignore
```

That the systemd named filedescriptor "service1" is attached to the socket matching on the second rule and the second named filedescriptor is then not matched anymore. It seems to me there might be an off-by-one indexing error in the logic handling systemd named filedescriptors. I'll provide some additional logs once I reproduce on a separate test system.

With only one named systemd filedescriptor things work out fine.

**Comment by aszlig** [comment]
> I'll provide some additional logs once I reproduce on a separate test system.

No need to, I could reproduce your issue in a small test, but so far it doesn't look like an indexing error:

<details>
<summary>NixOS VM test</summary>

```nix
import <nixpkgs/nixos/tests/make-test-python.nix> {
  name = "ip2unix-systemd";

  machine = { pkgs, lib, ... }: let
    ip2unix = import ../.. { inherit pkgs lib; };

    testServer = pkgs.writeScript "test-server.py" ''
      #!${pkgs.python3.interpreter}
      from functools import partial
      from selectors import DefaultSelector, EVENT_READ
      from http.server import BaseHTTPRequestHandler, HTTPServer

      class TestHandler(BaseHTTPRequestHandler):
        def __init__(self, *args, reply, **kwargs):
          self.reply = reply
          super().__init__(*args, **kwargs)

        def do_GET(self):
          self.send_response(200)
          self.send_header('Content-Type', 'text/plain')
          self.send_header('X-Reply', self.reply)
          self.end_headers()
          self.wfile.write(self.reply.encode())

      selector = DefaultSelector()
      for port in [1113, 8090]:
        handler = partial(TestHandler, reply=f'port{port}')
        server = HTTPServer(('127.0.0.1', port), handler)
        selector.register(server, EVENT_READ, server.handle_request)

      while True:
        for event in selector.select():
          event[0].data()
    '';

  in {
    systemd.sockets.testsock1 = {
      description = "Test Socket 1 for Multiple Socket Test";
      wantedBy = [ "sockets.target" ];
      requiredBy = [ "socktest.service" ];
      socketConfig.ListenStream = "/run/test1.sock";
      socketConfig.FileDescriptorName = "socket1";
      socketConfig.Service = "socktest.service";
    };

    systemd.sockets.testsock2 = {
      description = "Test Socket 2 for Multiple Socket Test";
      wantedBy = [ "sockets.target" ];
      requiredBy = [ "socktest.service" ];
      socketConfig.ListenStream = "/run/test2.sock";
      socketConfig.FileDescriptorName = "socket2";
      socketConfig.Service = "socktest.service";
    };

    systemd.services.canary = {
      description = "Canary Service to verify implementation";
      after = [ "network.target" ];
      requiredBy = [ "multi-user.target" ];
      serviceConfig.User = "testuser";
      serviceConfig.Group = "testgroup";
      serviceConfig.ExecStart = testServer;
    };

    systemd.services.socktest = {
      description = "Test Service for Multiple Socket Test";
      serviceConfig.User = "testuser";
      serviceConfig.Group = "testgroup";
      serviceConfig.ExecStart = lib.escapeShellArgs [
        "${ip2unix}/bin/ip2unix"
        "-r" "in,tcp,port=8090,systemd=socket1"
        "-r" "in,tcp,port=1113,systemd=socket2"
        "-r" "out,ignore"
        testServer
      ];
    };

    users.users.testuser.isSystemUser = true;
    users.users.testuser.group = "testgroup";
    users.groups.testgroup = {};
  };

  testScript = { nodes, ... }: let
    inherit (nodes.client.config.networking) primaryIPAddress;
  in ''
    machine.wait_for_unit('multi-user.target')

    for n, port in enumerate([8090, 1113], start=1):
      for args in [f'http://127.0.0.1:{port}/',
                   f'--unix-socket /run/test{n}.sock http://test/']:
        cmd = f'test "$(curl --no-progress-meter -v {args})" = port{port}'
        machine.succeed(cmd)
  '';
}
```
</details>

When adding `-vvvvv` to the `ip2unix` call, we get:

```
ip2unix:systemd.cc:223:init Matched systemd unix file descriptor name 'socket1' (fd 3) with rule #0.
ip2unix:systemd.cc:223:init Matched systemd unix file descriptor name 'socket2' (fd 4) with rule #1.
```

Rule `#0` maps to port 8090, which in turn has a `FileDescriptorName` of `socket1`, so this part looks correct.

However, where things go bonkers is a bit later:

```
ip2unix:preload.cc:252:operator() Systemd file descriptor queue empty, blackholing socket with fd 3.
ip2unix:socket.cc:97:blackhole Socket with fd 3 blackholed.
```

**Comment by aszlig** [comment]
Okay, turns out this actually is a regression and I've bisected the issue to commit 1a8a9b6ca73865a10452eb10c93dd751d0ba6af7.

**Comment by aszlig** [comment]
@noc0lour: Commit 5ead045bffe781e1133d3881d70a4b5bdc66021d should fix the issue. Can you please test the commit in question and check whether it's working for you?

**Comment by aszlig** [comment]
Closing this since the proposed fix is part of version 2.1.4. If the problem persist, please reopen or create a new issue. Thanks :-)
